### PR TITLE
ADR23 - Trace evaluation

### LINKED ISSUE #2188: Write ADR for trace evaluation
Title

### PR DESCRIPTION
**Pull request by Kukovec**
- ~[ ] Tests added for any new code~
- ~[ ] Ran `make fmt-fix` (or had formatting run automatically on all files edited)~
- ~[ ] Documentation added for any new functionality~
- ~[ ] [Entries added to `./unreleased/`][changelog format] for any new functionality~

[changelog format]: https://github.com/informalsystems/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

closes #2188

### CODE REVIEW

File: docs/src/adr/023adr-trace-evaluation.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+to achieve a better user experience\
+accepting the development costs.\
+
+## Context
+
+<!-- Communicates the forces at play (technical, political, social, project).
+     This is the story explaining the problem we are looking to resolve.
+-->
+As explained in [#1845](https://github.com/informalsystems/apalache/issues/1845), one often runs into the problem of unreadable counterexamples; 
+for complex specifications, it is often the case that either the sate contains many variables, or the invariant contains many conjunctive components (or both).
```
Comment (konnov) [inline docs/src/adr/023adr-trace-evaluation.md:31]:
```suggestion
for complex specifications, it is often the case that either the state contains many variables, or the invariant contains many conjunctive components (or both).
```

File: docs/src/adr/023adr-trace-evaluation.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+     problem. -->
+
+We choose option (2). 
+
+Suppose we are given a trace `s0 -> s1 -> ... -> sn` over variables `x1,...,xk` as well as expressions `E1,...,Em`, such that all free variables of `E1,...,Em` are among `x1,...,xk`. W.l.o.g. assume all constants are inlined.
+
+The above defines a trace `t0 -> t1 -> ... -> tn` over variables `v1,...,vm`, such that 
+```
+ti.vj = Ej[si.x1/x1,...,si.xk/xk]
+```
```
Comment (konnov) [inline docs/src/adr/023adr-trace-evaluation.md:60]: Could you use underscores in the variable names, e.g., `t_i`, `v_j`. It's hard to parse the current write up.

File: docs/src/adr/023adr-trace-evaluation.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+for all `i \in 0..n, j \in 1..m`. In other words, `vj` in state `ti` is the evaluation of the expression `Ej` in state `si`.
+
+By using TransitionExecutor instead of a generic Next-decomposition, this can be encoded as a specification free of control-nondeterminism, in-state computation, or invariants, and is thus incredibly efficient to represent in SMT.
```
Comment (konnov) [inline docs/src/adr/023adr-trace-evaluation.md:64]: The transition executor is very much in the guts of the model checker. While it is possible to use it, I do not think it is the simplest way of solving the problem.

Why cannot we do the following. Assume we have the following trace:

```tla
State0 ==
  /\ x = 2
  /\ y = 3

State1 ==
  /\ x = 4
  /\ y = 5

State2 ==
  /\ x = 6
  /\ y = 4
```

We could encode the above trace as a sequence of records:

```tla
Trace == <<
  [ x |-> 2, y |-> 3],
  [ x |-> 4, y |-> 5],
  [ x |-> 6, y |-> 4]
>>
```

Next, suppose we want to compute the value of `x + y` in every state of the trace. This could be done by producing the following spec:

```tla
\* the trace as defined above
Trace == << ... >>

VARIABLE
  \* the result of evaluating x + y
  \* @type: Int;
  eval1,
  \* The state number in the trace
  \* @type: Int;
  index

Init ==
  /\ eval1 = Trace[1].x + Trace[1].y
  /\ stateNo = 1

Next ==
  LET nextIndex == index + 1 IN
  /\ index' = nextIndex
  /\ eval1 = Trace[nextIndex].x + Trace[nextIndex].y
```

Note that the above specification can be produced by a preprocessing step. There is no need for direct interaction with the transition executor.

**Review by konnov — COMMENTED** [review]
It is clear to me that we can solve the problem in many ways. I would prefer to solve it by pure preprocessing (see my comment). Also, this ADR is written in a rush. Could you add an example that illustrates the input and the output. In this case, it should be really easy to do that.

File: docs/src/adr/023adr-trace-evaluation.md (hunk 3 above)
Comment (Kukovec) [reply docs/src/adr/023adr-trace-evaluation.md:64]: You could do that, but then you're actually solving additional sequence and record applications in SMT. The fact that the executor is "in the guts of the model checker" is irrelevant; the user doesn't need to know or care. 
If your claim is that the executor is too difficult for Apalache developers to use, however, that is an entirely separate discussion.

File: docs/src/adr/023adr-trace-evaluation.md (hunk 2 above)
Comment (Kukovec) [reply docs/src/adr/023adr-trace-evaluation.md:60]: I can, but I don't think markdown prettyprints those.

File: docs/src/adr/023adr-trace-evaluation.md (hunk 3 above)
Comment (konnov) [reply docs/src/adr/023adr-trace-evaluation.md:64]: > If your claim is that the executor is too difficult for Apalache developers to use, however, that is an entirely separate discussion.

Why? This is an ADR. We can discuss the implications of development costs and long-term support :)

Comment (Kukovec) [reply docs/src/adr/023adr-trace-evaluation.md:64]: > Why? This is an ADR. We can discuss the implications of development costs and long-term support :)

I think you have to write your ADRs either assuming the rest of the code is what you've got, OR, write an ADR on how to refactor/remove TEX _first_. Otherwise you get scope creep and your ADRs inevitably become a series of cascading changes and refactoring.

Comment (Kukovec) [reply docs/src/adr/023adr-trace-evaluation.md:64]: I've explored it a bit, and I suppose we can keep this high level by using transition filtering instead:
```
0->0|1->1|...|n->n
```
should give us the i-th Next in the i-th step while allowing us to just call the model checker pass w/o disassembling it.

File: docs/src/adr/023adr-trace-evaluation.md (hunk 2 above)
Comment (shonfeder) [reply docs/src/adr/023adr-trace-evaluation.md:60]: I think you can use latex, in case that would help: https://github.blog/2022-05-19-math-support-in-markdown/

**Review by shonfeder — COMMENTED** [review]
This proposal looks pretty sensible to me and I think I can understand the motivation.

It's not clear to me exactly what the tradeoffs are between Jure's proposed approach and what Igor is proposing, but it seems like it would make sense to incorporate the alternative Igor is suggesting and add some justification for why we do or do not take that proposed route.

File: docs/src/adr/023adr-trace-evaluation.md (hunk 3 above)
Comment (Kukovec) [reply docs/src/adr/023adr-trace-evaluation.md:64]: If you're using trace exploration, user-supplied transition filtering makes no sense; there is exactly one transition between each two given states on a trace.

Comment (thpani) [reply docs/src/adr/023adr-trace-evaluation.md:64]: > user-supplied transition filtering makes no sense

Any other tuning option then. Currently, tuning options are just a serialized String->String map.

Comment (Kukovec) [reply docs/src/adr/023adr-trace-evaluation.md:64]: Trace evaluation isn't really typical BMC (it just uses it as a backend). The user should not be providing any additional parameters at all, besides the trace file and a list of expressions.

Comment (thpani) [reply docs/src/adr/023adr-trace-evaluation.md:64]: Maybe you can explain what "_By using TransitionExecutor instead of a generic Next-decomposition, this can be encoded as a specification_" actually means.
It would be good to include such a specification, as an example. It's hard to argue about things that aren't explained in the ADR.

Comment (Kukovec) [reply docs/src/adr/023adr-trace-evaluation.md:64]: I actually have a prototype, and I'll demo it tomorrow. Afterwards, we can update the ADR.

Comment (Kukovec) [reply docs/src/adr/023adr-trace-evaluation.md:64]: Also, this design has been discussed offline for at least 2-3 months before the ADR.

Comment (Kukovec) [reply docs/src/adr/023adr-trace-evaluation.md:64]: I think a lot of the discussion happening here has nothing to do with _design_, but is nitpicking implementation details.

Comment (thpani) [reply docs/src/adr/023adr-trace-evaluation.md:64]: > the thing being discussed in a very specific implementation detail, and the prototype has uncontroversial stuff

Then demoing the prototype won't change the discussion here.

Comment (shonfeder) [reply docs/src/adr/023adr-trace-evaluation.md:64]: I also think it's worth keeping in mind that our use of ADRs are kind of overloaded. iiuc, they were principally intended to be architectural **decision** records: and in that function it makes sense to me we may want a light-weight prototype to help inform our decisions. But we have also often used them as design documents, and perhaps this dual use can cause some confusion w/r/t expectations?

Comment (Kukovec) [reply docs/src/adr/023adr-trace-evaluation.md:64]: > > Also, this design has been discussed offline for at least 2-3 months before the ADR.
> 
> Apparently not to a point where there's consensus on how to steer the transition exploration. Documenting all options on the table and weighing their benefits is exactly the point of an ADR.

But there _is_ consensus on how to steer transition exploration. Take `Next0` in the initial state, then `Next1`, etc.

The different options are about facilitating this choice of transitions (either via TRex calls or via tuning parameters, which get internally propagated into TRex calls).

File: docs/src/adr/023adr-trace-evaluation.md (hunk 2 above)
Comment (konnov) [reply docs/src/adr/023adr-trace-evaluation.md:60]: > I think you can use latex, in case that would help: https://github.blog/2022-05-19-math-support-in-markdown/

Yeah, that's a very cool feature. But the ADRs are going via mdbook, so this would not work I suppose? We had some other special support for math environment.

Comment (Kukovec) [reply docs/src/adr/023adr-trace-evaluation.md:60]: I used some mdbook-latex library in https://apalache.informal.systems/docs/tutorials/symbmc.html.
In hindsight, probably a massive waste of time, since I reckon like 3 people ever read that.

File: docs/src/adr/023adr-trace-evaluation.md (hunk 3 above)
Comment (konnov) [reply docs/src/adr/023adr-trace-evaluation.md:64]: > The different options are about facilitating this choice of transitions (either via TRex calls or via tuning parameters, which get internally propagated into TRex calls).

Or via preprocessing without using TRex ;-)

Comment (Kukovec) [reply docs/src/adr/023adr-trace-evaluation.md:64]: > Or via preprocessing without using TRex ;-)

Name one advantage this has over the solution with `tuning`.

Comment (Kukovec) [reply docs/src/adr/023adr-trace-evaluation.md:64]: Plus, `Trace[nextIndex]` dodges all preprocessing and, most importantly, const simplification.

Comment (thpani) [reply docs/src/adr/023adr-trace-evaluation.md:64]: > they were principally intended to be architectural **decision** records

Sure, but where do we discuss a proposed decision, if not on the PR merging the decision record?

> in that function it makes sense to me we may want a light-weight prototype to help inform our decisions

Sure, as long as such a prototype informs decisions and does not preempt them :)

Comment (Kukovec) [reply docs/src/adr/023adr-trace-evaluation.md:64]: > Of course, as long as such a prototype informs decisions and does not preempt them :)

Pretty much the only tradeoff/downside of my proposed approach is _potential_ dev/maintenance cost. I find it pointless to argue how complicated an abstract implementation would be to maintain/extend. If you have a prototype in front of you, you can start making claims about which parts are maintainable and which aren't.

**Comment by thpani** [comment]
@Kukovec I think you might've pushed the wrong branch, or some merge commit crept in here?

**Comment by Kukovec** [comment]
> @Kukovec I think you might've pushed the wrong branch, or some merge commit crept in here?

Huh, I only did `rebase origin/main`, but let me look at that again.

**Comment by Kukovec** [comment]
Yeah, guess it was just that. Thanks for noticing.

File: docs/src/adr/023adr-trace-evaluation.md [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+(* Transition 3 to State4 *)
+State4 == A = 25/\ B = 1/\ C = {}/\ D = FALSE
+
+(* The following formula holds true in the last state and violates the invariant *)
+InvariantViolation == TRUE
+
+================================================================================
+(* Created by Apalache on Mon Oct 17 16:06:59 CEST 2022 *)
+(* https://github.com/informalsystems/apalache *)
+```
```
Comment (shonfeder) [inline docs/src/adr/023adr-trace-evaluation.md:401]: Thanks! do find having the example here really helps.

After our discussion today, I think it would be worth including a note about how operators can be easily added to get views into regions of the state one may be interested in.

Also, I do think it's worth considering the interaction with the trace viewer, and I do think @thpani makes a strong case here.

IMO, it's fine if we decide to not pursue the course he advises (to extend the state with these  evaluations, and leave filtering to the viewer), but the rationale for that choice is probably good to record here (perhaps it's just because it takes less work! :) .

With those additions, for my part, I'm quite happy to have this merged in.

**Review by shonfeder — APPROVED** [review]

File: docs/src/adr/023adr-trace-evaluation.md [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+
+## Summary
+
+<!-- Statement to summarize, following the following formula: -->
+
+In the context of improving usability\
+facing difficulties understanding counterexample traces\
+we decided to implement trace evaluation\
+to achieve a better user experience\
+accepting the development costs.\
```
Comment (thpani) [inline docs/src/adr/023adr-trace-evaluation.md:23]: Somehow github renders this as literal `\`:

```suggestion
accepting the development costs.
```

**Review by thpani — APPROVED** [review]
LGTM, thanks for extending the document!

I've noted my preference for output option (1), because I think it's a more universally usable output format. That said, I'm fine going either way and if necessary iterating upon UX feedback.

File: docs/src/adr/023adr-trace-evaluation.md (hunk 4 above)
Comment (thpani) [reply docs/src/adr/023adr-trace-evaluation.md:401]: That's a good point, maybe we can include a link to https://github.com/informalsystems/vscode-itf-trace-viewer and note that it supports filtering by variable?
